velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -110,6 +110,9 @@ class Addition {
     r.setUnscaledValue(
         a.unscaledValue() * DecimalUtil::kPowersOfTen[aRescale] +
         b.unscaledValue() * DecimalUtil::kPowersOfTen[bRescale]);
+    if (!UnscaledLongDecimal::valueInRange(r.unscaledValue())) {
```

Comment:
I think we need to re-write the template as
```
r = checkedAdd<R>(
              checkedMultiply<R>(a, A(DecimalUtil::kPowersOfTen[aRescale])), 
              checkedMultiply<R>(b, B(DecimalUtil::kPowersOfTen[bRescale]))
          );
```
What do you think?

Reply:
We should remove `setUnscaledValue` API since we are not checking bounds there. I don't think we need this API anyway.

Reply:
Writing template this way may not be right. A is UnscaledShortDecimal and kPowersOfTen[aRescale] has greater thank 18 precision, it will lead to incorrect values.
```
 checkedMultiply<R>(a, A(DecimalUtil::kPowersOfTen[aRescale])),
```

I introduced a new template under CheckedArithmetic.h. Please take a look.

Reply:
I think the template adds additional complexity. What about?
```
r = checkedAdd<R>(
              checkedMultiply<R>(R(a), R(DecimalUtil::kPowersOfTen[aRescale])), 
              checkedMultiply<R>(R(b), R(DecimalUtil::kPowersOfTen[bRescale]))
          );
```